pkg/unshare/unshare_linux.go

```diff
@@ -387,10 +388,42 @@ const (
 	UsernsEnvName = "_CONTAINERS_USERNS_CONFIGURED"
 )
 
+// hasFullUsersMappings checks whether the current user namespace has all the IDs mapped.
+// current user namespace
+func hasFullUsersMappings() (bool, error) {
+	content, err := ioutil.ReadFile("/proc/self/uid_map")
+	if err != nil {
+		return false, err
+	}
+	// Both runc and crun check whether the current process is in a user namespace
+	// by looking up 4294967295 in /proc/self/uid_map.  If the mappings would be
+	// copied as they are, the check in the OCI runtimes would fail.  So just split
+	// it in two different ranges.
+	return bytes.Contains(content, []byte("4294967295")), nil
+}
+
 // IsRootless tells us if we are running in rootless mode
 func IsRootless() bool {
 	isRootlessOnce.Do(func() {
 		isRootless = getRootlessUID() != 0 || getenv(UsernsEnvName) != ""
+		if !isRootless {
+			hasCapSysAdmin, err := HasCapSysAdmin()
+			if err != nil {
+				logrus.Warnf("Failed to read CAP_SYS_ADMIN presence for the current process")
+			}
+			if err == nil && !hasCapSysAdmin {
+				isRootless = true
+			}
+		}
+		if !isRootless {
+			hasMappings, err := hasFullUsersMappings()
```

Comment:
Why would this indicate isRootless.

If I have a range of `0 0 1000000` won't the isRootless return true?  Aren't I running as root in this case?

Reply:
But we are using `isRootless` to mean "not in the initial user namespace".  That is what the kernel does when checking what is allowed and what not.  For example, we need to use `user. ` xattrs for overlay whiteouts even if we are root in a user namespace

Reply:
I think by now, we **desperately need** documentation. “IsRootless tells us if we are running in rootless mode” says nothing.

Describe in which situations that function returns true (possibly how can one get into such situations), and what does that means for application behavior. What should, or shouldn’t, they be doing differently?

It might _well be the case_ that we need three or more different functions, `IsNotInTheInitialUserNamespace`, `IsInsideAPodmanCreatedContainerNamespace`,  and so on.

Just vaguely saying “rootless” is apparently causing misunderstandings and unanticipated breakage.